hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -645,15 +604,50 @@ public void run() {
             }
 
             try {
-                authenticate(target, connection, asOwner, future);
+                authenticateAsync();
             } catch (Exception e) {
                 future.onFailure(e);
                 connection.close("Failed to authenticate connection", e);
                 connectionsInProgress.remove(target);
             }
         }
 
-        private ClientConnection getConnection(Address target) throws IOException {
+        private void authenticateAsync() {
+            this.principal = getPrincipal();
```

Comment:
Getting principal can be moved to `encodeAuthenticationRequest`. 
No need for it to be a field on the class.